mod/beacon/blockchain/service.go

```diff
@@ -164,134 +166,154 @@ func (s *Service[
+	}
+
+	if err := s.dispatcher.Subscribe(
+		events.FinalBeaconBlockReceived, s.subFinalBlkReceived,
+	); err != nil {
 		return err
 	}
-	go s.start(ctx, subBlkCh, subGenCh)
+
+	// start a goroutine to listen for requests and handle accordingly
+	go s.eventLoop(ctx)
 	return nil
 }
 
+// eventLoop listens for events and handles them accordingly.
 func (s *Service[
 	_, BeaconBlockT, _, _, _, _, _, _, GenesisT, _,
-]) start(
-	ctx context.Context,
-	subBlkCh chan *asynctypes.Event[BeaconBlockT],
-	subGenCh chan *asynctypes.Event[GenesisT],
-) {
+]) eventLoop(ctx context.Context) {
 	for {
 		select {
 		case <-ctx.Done():
 			return
-		case msg := <-subBlkCh:
-			switch msg.Type() {
-			case events.BeaconBlockReceived:
-				s.handleBeaconBlockReceived(msg)
-			case events.BeaconBlockFinalizedRequest:
-				s.handleBeaconBlockFinalization(msg)
-			}
-		case msg := <-subGenCh:
-			if msg.Type() == events.GenesisDataProcessRequest {
-				s.handleProcessGenesisDataRequest(msg)
-			}
+		case event := <-s.subGenDataReceived:
+			s.handleProcessGenesisDataEvent(event)
+		case event := <-s.subBlockReceived:
+			s.handleVerifyBeaconBlockEvent(event)
+		case event := <-s.subFinalBlkReceived:
+			s.handleFinalizeBeaconBlockEvent(event)
 		}
 	}
 }
 
+/* -------------------------------------------------------------------------- */
+/*                              Message Handlers                              */
+/* -------------------------------------------------------------------------- */
+
 func (s *Service[
 	_, _, _, _, _, _, _, _, GenesisT, _,
-]) handleProcessGenesisDataRequest(msg *asynctypes.Event[GenesisT]) {
+]) handleProcessGenesisDataEvent(msg async.Event[GenesisT]) {
+	var (
+		valUpdates transition.ValidatorUpdates
+		genesisErr error
+	)
 	if msg.Error() != nil {
 		s.logger.Error("Error processing genesis data", "error", msg.Error())
-		return
 	}
 
 	// Process the genesis data.
-	valUpdates, err := s.ProcessGenesisData(msg.Context(), msg.Data())
-	if err != nil {
-		s.logger.Error("Failed to process genesis data", "error", err)
+	valUpdates, genesisErr = s.ProcessGenesisData(msg.Context(), msg.Data())
+	if genesisErr != nil {
+		s.logger.Error("Failed to process genesis data", "error", genesisErr)
 	}
 
-	// Publish the validator set updated event.
-	if err = s.validatorUpdateBroker.Publish(
-		msg.Context(),
-		asynctypes.NewEvent(
+	// Emit the event containing the validator updates.
+	if err := s.dispatcher.Publish(
+		async.NewEvent(
 			msg.Context(),
-			events.ValidatorSetUpdated,
+			events.GenesisDataProcessed,
 			valUpdates,
-			err,
+			genesisErr,
 		),
 	); err != nil {
 		s.logger.Error(
-			"Failed to publish validator set updated event",
-			"error",
-			err,
+			"Failed to emit event in process genesis data",
+			"error", err,
 		)
+		panic(err)
 	}
 }
 
 func (s *Service[
 	_, BeaconBlockT, _, _, _, _, _, _, _, _,
-]) handleBeaconBlockReceived(
-	msg *asynctypes.Event[BeaconBlockT],
+]) handleVerifyBeaconBlockEvent(
```

Comment:
this should still be received imo, bc ignoring comet, this is the first time this node as seen the block. We are verifiying the block as a result of receving it, it shouldnt be labelled as processing a request.

Reply:
yeah agree